modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/DropZone.es.js

```diff
@@ -0,0 +1,140 @@
+import React, {Component} from 'react';
+import PropTypes from 'prop-types';
+import {DropTarget as dropTarget} from 'react-dnd';
+import {DragTypes} from '../../utils/drag-types.es';
+import getCN from 'classnames';
+
+const {CRITERIA_GROUP, CRITERIA_ROW, PROPERTY} = DragTypes;
+
+const acceptedDragTypes = [
+	CRITERIA_GROUP,
+	CRITERIA_ROW,
+	PROPERTY
+];
+
+/**
+ * Prevents groups from dropping within itself and all items from dropping into
+ * a position that would not change its' current position.
+ * This method must be called `canDrop`.
+ * @param {Object} props Component's current props.
+ * @param {DropTargetMonitor} monitor
+ * @returns {boolean} True if the target should accept the item.
+ */
+function canDrop(props, monitor) {
+	const {
+		groupId: destGroupId,
+		index: destIndex
+	} = props;
+
+	const {
+		childGroupIds = [],
+		criterion,
+		groupId: startGroupId,
+		index: startIndex
+	} = monitor.getItem();
+
+	const disallowedGroupIds = [criterion.groupId, ...childGroupIds];
+
+	const sameOrNestedGroup = monitor.getItemType() === CRITERIA_GROUP &&
+		disallowedGroupIds.includes(destGroupId);
+
+	const sameIndexInSameGroup = startGroupId === destGroupId &&
+		(startIndex === destIndex || startIndex === destIndex - 1);
+
+	let droppable = true;
+
+	if (sameOrNestedGroup || sameIndexInSameGroup) {
+		droppable = false;
+	}
+
+	return droppable;
```

Comment:
I'd probably use a ternary here rather than the "if" block above. Makes it a little more concise an doesn't hurt readability.